### ISSUE TITLE
Segfault due to string interpolation in a method

### ISSUE THREAD

**Issue by quinnj** [JuliaLang/julia#6387]
Reported at github.com/karbarcca/Datetime.jl#42.
Reduced case:
```julia
using Datetime
 
typealias DateRanges Union(DateRange,DateRange1)
 
type ObjMember
  member::DateRanges
end

MemorialDay = x->month(x)==May && dayofweek(x)==Monday && dayofweekinmonth(x)==daysofweekinmonth(x)
day = date(2014, 04, 01)
obj = ObjMember(date(2014):MemorialDay:date(2015))
interval = [obj.member]
# Don't segfault
day in interval
println("day_in(day::Date, obj::ObjMember) <= $(typeof(day)), $(typeof(obj))($(typeof(obj.member)))")

function day_in(day::Date, obj::ObjMember)
  println("day_in(day::Date, obj::ObjMember) <= $(typeof(day)), $(typeof(obj))($(typeof(obj.member)))")
  interval = [obj.member] # cool
  return (day in interval) # not cool
end
# Does segfault
day_in(day,obj)
```

**Comment by timholy** [comment]
I'm reopening this. In the context of #10911, I'm struggling with this one. Eventually I had the "brilliant" idea of testing the exact same thing on current master:
```jl
$ gdb ./julia
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
[snip]
(gdb) run
Starting program: /home/tim/src/julia/julia 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffdef3fa700 (LWP 10965)]
[New Thread 0x7ffdeebf9700 (LWP 10966)]
[New Thread 0x7ffdec3f8700 (LWP 10967)]
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+4572 (2015-04-29 20:24 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit dea3d0e* (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> bitstype 64 Date6387{C}

julia> type DateRange6387{C} <: Range{Date6387{C}}
              end

julia> DateRange6387{Int64}()

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7e92b48 in ?? ()
(gdb)
```
This happens despite the fact that `make test-core` passes for me.

**Comment by timholy** [comment]
Ah, the segfault does not happen until one tries to show the value; `obj = DateRange6387{Int64}();` does not segfault, but then typing `obj` produces a stack overflow.

I can reproduce the segfault with this script:
```jl
bitstype 64 Date6387{C}
type DateRange6387{C} <: Range{Date6387{C}}
end
obj = DateRange6387{Int64}()
iob = IOBuffer()
show(iob, obj)
nothing
```
or insert similar lines in `test/core.jl` and get it too. (So it doesn't look like some awful history-dependent problem.)

**Comment by timholy** [comment]
Never mind; it comes down to circular definitions [here](https://github.com/JuliaLang/julia/blob/dea3d0e42029af05f58a0069491238462382e591/base/abstractarray.jl#L65) and [here](https://github.com/JuliaLang/julia/blob/dea3d0e42029af05f58a0069491238462382e591/base/range.jl#L347-L350), given that this type does not yet have `first` defined. Sigh. Sorry for the noise.

**Comment by pao** [comment]
@timholy Despite that I don't see any constructions that should be permitted to segfault. Forgetting a part of the totally implicit, unchecked, and not terribly well documented interface shouldn't crash, it should help you figure out what isn't implemented yet (though possibly as a new issue).

**Comment by ScottPJones** [comment]
+100!

**Comment by timholy** [comment]
No, in this case I was just being dumb. It's a totally-expected StackOverflow error, which is what you expect from the equivalent of `foo() = foo()`. That's indeed exactly the error that it gives when you're not running under the debugger, but we catch StackOverflows via [a segfault mechanism](https://github.com/JuliaLang/julia/blob/dea3d0e42029af05f58a0069491238462382e591/src/init.c#L272-L277), and for #10911 I'm living in the debugger which intercepts the original segfault.

**Comment by timholy** [comment]
But for posterity, it may be worth explaining how I got confused, since someone else may someday benefit from my experience:
- The problem: you're monkeying with julia's internals and trying to get your code working. You try one of julia's test files, and you get a segfault. There's no indication where in this 2000-line test file your segfault was triggered.
- Not-so-bright idea: "I know, I'll just copy-paste the contents into the REPL---then I'll _see_ which line triggered the segfault!"
- You get a segfault on line 1234. You think, "yay, I've found the origin of the segfault!"
- The problem with this idea: some of our tests have not been designed to support calling `show` on the output, but when you copy/paste into the REPL, suddenly all the outputs are `show`n. This can mislead you into thinking that something is broken when it is not---in this case, a StackOverflow for a type whose incomplete "interface" would be a problem if it were a "serious" definition, but which is perfectly acceptable if all you're doing is defining the type to test a specific compiler/type intersection/etc. issue.
- A better approach: run the tests but don't show the results. The following function seems to work well, and allowed me to discover that my segfault happens at a much later point in the same file:
```jl
function test_showline(filename)
    str = readall(filename)
    eval(Main, parse("using Base.Test"))
    idx = 1
    while idx < length(str)
        ex, idx = parse(str, idx)
        @show idx
        @show ex
        eval(Main,ex)
    end
    println("done")
end
```
